src/codegen.cpp

```diff
@@ -188,6 +188,41 @@ void setName(jl_codegen_params_t &params, Value *V, std::function<std::string()>
     }
 }
 
+void setNameWithField(jl_codegen_params_t &params, Value *V, std::function<StringRef()> GetObjName, jl_datatype_t *jt, unsigned idx, const Twine &suffix)
+{
+    assert((isa<Constant>(V) || isa<Instruction>(V)) && "Should only set names on instructions!");
+
+    if (jl_is_tuple_type(jt)){
+        if (params.debug_level >= 2 && !isa<Constant>(V)) {
+            V->setName(Twine(GetObjName()) + "[" + Twine(idx + 1) + "]"+ suffix);
+        }
+        return;
+    }
+
+    const char *field;
+    if (jl_is_namedtuple_type(jt)) {
+        auto names = jl_tparam0(jt);
+        assert(jl_is_tuple(names));
+        if (idx < jl_nfields(names)) {
+            auto name = jl_fieldref(names, idx);
+            assert(jl_is_symbol(name));
+            field = jl_symbol_name((jl_sym_t*)name);
+        }
+    } else {
+        auto flds = jl_field_names(jt);
+        if (idx < jl_svec_len(flds)) {
+            auto name = jl_svec_ref(flds, idx);
+            assert(jl_is_symbol(name));
+            field = jl_symbol_name((jl_sym_t*)name);
+        }
+    }
+    field = "<unknown field>";
+    assert((isa<Constant>(V) || isa<Instruction>(V)) && "Should only set names on instructions!");
+    if (params.debug_level >= 2 && !isa<Constant>(V)) {
+        V->setName(Twine(GetObjName()) + "." + Twine(field) + suffix);
+    }
+}
```

Comment:
Now that I see my old code, I see that I probably defined the lazy setName API wrongly; I should have made it something like
```c++
static void setLazyName(jl_codegen_params_t &emission_context, Value *V, llvm::function_ref<void(Value *)> NameSetter)
```

This way you don't need to construct that std::string in the middle, and it respects the Twine lifetime considerations. We could also make the `function_ref` a template, though that decreases readability for probably not a big performance win.

Reply:
The actual problem here is that you compute the fieldname even when we don't actually set it on the value; that would be easier to avoid with the above API, since it's obvious which code runs only if we're actually using the Twine.

Reply:
Oh, I just need to move that if debug_level to above all the work, not sure why I didn't do that.